python/lsst_versions/_versions.py

```diff
@@ -0,0 +1,246 @@
+                continue
+            # Assume that only major releases matter when looking through
+            # the history.
+            major = tag_name[1:].split(".")[0]
+            if major:
+                releases[int(major)] = release.object
+        elif tag_name.startswith("w."):
+            weekly = tagref.tag
+            if weekly is None:
+                continue
+
+            # There can be multiple weeklies associated with a single
+            # commit. Retain the newest weekly. Some weekly tags did not
+            # zero pad the week so must be normalized before comparison.
+            if len(tag_name) == 8:
+                y, w = tuple(int(i) for i in tag_name.split(".")[1:])
+                tag_name = f"w.{y}.{w:02d}"
+
+            # Store the weeklies associated with the object they are tagging
+            # but only if this weekly is more recent than the one that may
+            # already be stored.
+            hexsha = weekly.object.hexsha
+            # print(f"Checking {tag_name} [{hexsha}]")
+            previous = None
+            if (previous := weeklies.get(hexsha, None)) and previous > tag_name:
+                continue
+            # print(f"Storing {hexsha} for tag {tag_name}")
+            weeklies[hexsha] = tag_name
+
+    commit = repo.commit(version_commit)
+
+    # Scan through all the releases for the first that does not have this
+    # commit as an ancestor.
+    relevant_release = 0
+    for major_release in sorted(releases, reverse=True):
+        if not repo.is_ancestor(commit, releases[major_release]):
+            relevant_release = major_release
+            break
+        if commit == releases[major_release]:
+            # It may be the actual release that should not be a dev
+            # release at all, but for now we do not try to look for
+            # release (non-rc) tags to deal with this properly.
+            relevant_release = major_release - 1
+            break
+
+    if relevant_release == 0:
+        log.warning("Could not find release tag in repo '%s', using 0.", repo_dir)
+
+    # Look through the parents until we find a weekly commit.
+    # The counter can report confusing results if this is being used for
+    # an unmerged development branch.
+    weekly_name, counter = _find_weekly_parent(0, commit, weeklies)
+    if not weekly_name:
+        # No weekly was found. This must be a very early commit.
+        year, week = "0", "0"
+    else:
+        year, week = weekly_name[2:].split(".")
+
+    # Python pre-release versions must can only have a single integer
+    # after the "a".
+    dev_version = f"{relevant_release + 1}.0.0a{int(year):04d}{int(week):02d}{counter:02d}"
+    log.debug(
+        "Using version %s for commit %s derived from weekly %s", dev_version, commit.hexsha, weekly_name
+    )
+
+    return dev_version
+
+
+def _load_toml(content: str) -> Dict[str, Any]:
+    """Parse the TOML contents using deferred import."""
```

Comment:
Mostly because that's how setuptools_scm set itself up so that the loader could be overridden (I think because they support other config formats). I suppose I'm only going to support TOML so I can move it down.

Reply:
Relocated.